server/autograder.py

```diff
@@ -105,3 +107,146 @@ def submit_continous(backup):
+    """
+    return send_autograder('/results', job_ids)
+
+GradingStatus = enum.Enum('GradingStatus', [
+    'QUEUED',   # a job is queued
+    'RUNNING',  # a job is running
+    'WAITING',  # the last job has finished, and we are waiting for a score
+    'DONE',     # we have a score
+    'FAILED',   # we could not get a score after several retries
+])
+
+class GradingTask:
+    def __init__(self, status, backup_id, job_id, retries):
+        self.status = status
+        self.backup_id = backup_id
+        self.job_id = job_id
+        self.retries = retries
+        self.status_change_time = time.time()
+
+    def set_status(self, status):
+        self.status = status
+        self.status_change_time = time.time()
+
+    def expired(self, timeout):
+        """Returns True if it has been at least TIMEOUT seconds since the last
+        status change.
+        """
+        return time.time() > self.status_change_time + timeout
+
+MAX_RETRIES = 3           # maximum number of times to retry a score
+QUEUED_TIMEOUT = 30 * 60  # maximum time or an autograder job to be queued for, in seconds
+RUNNING_TIMEOUT = 5 * 60  # time to wait for an autograder job to run, in seconds
+WAITING_TIMEOUT = 2 * 60  # time to wait for a score, in seconds
+POLL_INTERVAL = 10        # how often to poll the autograder, in seconds
+
+@jobs.background_job
+def autograde_assignment(assignment_id):
+    """Autograde all enrolled students for this assignment.
+
+    We set up a state machine for each backup to check its progress through
+    the autograder. If any step takes too long, we'll retry autograding that
+    backup. Ultimately, a backup is considered done when we confirm that
+    we've received a new score, or if we have reached the retry limit.
+    """
+    logger = jobs.get_job_logger()
+
+    assignment = Assignment.query.get(assignment_id)
+    course_submissions = assignment.course_submissions(include_empty=False)
+    backup_ids = set(fs['backup']['id'] for fs in course_submissions if fs['backup'])
+    token = create_autograder_token(jobs.get_current_job().user_id)
+
+    # start by sending a batch of all backups
+    start_time = time.time()
+    job_ids = send_batch(token, assignment, backup_ids)
+    tasks = [
+        GradingTask(
+            status=GradingStatus.QUEUED,
+            backup_id=backup_id,
+            job_id=job_id,
+            retries=0,
+        )
+        for backup_id, job_id in job_ids.items()
+    ]
+    num_tasks = len(tasks)
+
+    def retry_task(task):
+        if task.retries >= MAX_RETRIES:
+            logger.error('Did not receive a score for backup {} after {} retries'.format(
+                utils.encode_id(task.backup_id), MAX_RETRIES))
+            task.set_status(GradingStatus.FAILED)
+        else:
+            task.set_status(GradingStatus.QUEUED)
+            task.job_id = autograde_backup(token, assignment, task.backup_id)
+            task.retries += 1
+
+    while True:
+        time.sleep(POLL_INTERVAL)
```

Comment:
I agree that the log is not very interactive, but since there's no way to kill a job I don't think getting immediate feedback would be that useful anyway. We already have a score plot in the works and I think that would be a better way to monitor the distribution of scores after the run completes. Should we link to it somehow?